File: hotel-casa-blanca-api/app/schemas/reserva.py
"""Schemas de reservas (API_REST.md, seccion 5.3)."""
from datetime import date, datetime

from pydantic import BaseModel, Field

from app.schemas.common import Monto, ORMModel


class ReservaCreate(BaseModel):
    """
    Lo unico que manda la app. `monto_total` y `codigo_reserva` NO se aceptan
    del cliente: los calcula y genera el backend.
    """

    id_habitacion: int
    fecha_ingreso: date
    fecha_salida: date
    cantidad_personas: int = Field(ge=1)


class HabitacionResumen(ORMModel):
    """Datos minimos para pintar la lista de reservas sin otra llamada."""

    numero_habitacion: str | None = None
    tipo_nombre: str | None = None
    imagen_url: str | None = None


class ReservaResponse(ORMModel):
    id_reserva: int
    codigo_reserva: str
    id_habitacion: int
    id_estado_reserva: int
    fecha_ingreso: date
    fecha_salida: date
    cantidad_personas: int
    monto_total: Monto
    fecha_reserva: datetime
    habitacion: HabitacionResumen | None = None

File: apphotelcasablanca/lib/features/reservas/domain/entities/reserva.dart
// UBICACIÓN: lib/features/reservas/domain/entities/reserva.dart
import 'package:equatable/equatable.dart';

import '../../../../core/enums/estado_reserva.dart';

/// Tabla `reserva`. Incluye algunos datos de la habitación (numero, tipo,
/// imagen) que el backend embebe para poder pintar la lista sin otra llamada.
class Reserva extends Equatable {
  const Reserva({
    required this.id,
    required this.codigo,
    required this.idHabitacion,
    required this.fechaIngreso,
    required this.fechaSalida,
    required this.cantidadPersonas,
    required this.montoTotal,
    required this.estado,
    this.fechaReserva,
    this.habitacionNumero,
    this.tipoNombre,
    this.imagenUrl,
  });

  final int id;
  final String codigo;
  final int idHabitacion;
  final DateTime fechaIngreso;
  final DateTime fechaSalida;
  final int cantidadPersonas;
  final double montoTotal;
  final EstadoReserva estado;
  final DateTime? fechaReserva;

  // Datos embebidos de la habitación (para la lista).
  final String? habitacionNumero;
  final String? tipoNombre;
  final String? imagenUrl;

  int get noches => fechaSalida.difference(fechaIngreso).inDays;

  @override
  List<Object?> get props => [id, codigo, estado, fechaIngreso, fechaSalida];
}